mlir/tools/miopen-gen/miopen-gen.cpp

```diff
@@ -803,60 +822,342 @@ static func::FuncOp getMemsetFunc(ModuleOp module, mlir::Type elemType) {
+    loopIVs.append("kyx");
+    break;
+  case miopen::ConvOpType::BwdWeight:
+    std::tie(std::ignore, out_h, out_w) =
+        getConv2dBounds(miopen::ConvOpType::BwdData, genConfig);
+    llvm::copy(genConfig.filterDimension, std::back_inserter(upperBounds));
+    upperBounds.push_back(dimX);
+    upperBounds.push_back(out_h);
+    upperBounds.push_back(out_w);
+    loopIVs.append(genConfig.filterLayout);
+    loopIVs.append("nhw");
+    break;
   }
-  func =
-      func::FuncOp::create(loc, funcName, b.getFunctionType(funcArgTypes, {}));
-  module.push_back(func);
 
-  // Construct a new Block.
+  auto createConv2dLoopNest = [elemType, heightMap, widthMap, dimH, dimW, block,
+                               loopIVs, &genConfig](OpBuilder b, Location loc,
+                                                    ValueRange ivs) {
+    auto strideHeight =
+        b.create<arith::ConstantIndexOp>(loc, genConfig.strideHeight);
+    auto strideWidth =
+        b.create<arith::ConstantIndexOp>(loc, genConfig.strideWidth);
+    mlir::Value heightIdx, widthIdx;
+    mlir::Value heightTempIdx, widthTempIdx;
+
+    switch (genConfig.operation.getValue()) {
+    case miopen::ConvOpType::Fwd:
+      // in_h_idx = out_h_idx * stride_h + fil_h_idx * dilation_h - padding_h_l;
+      // in_w_idx = out_w_idx * stride_w + fil_w_idx * dilation_w - padding_w_l;
+      heightIdx = b.create<AffineApplyOp>(
+          loc, heightMap,
+          ValueRange{ivs[genConfig.outputLayout.find('h')], ivs[6]});
+      widthIdx = b.create<AffineApplyOp>(
+          loc, widthMap,
+          ValueRange{ivs[genConfig.outputLayout.find('w')], ivs[7]});
+      break;
+    case miopen::ConvOpType::BwdData:
+      // out_h_tmp = in_h_idx + padding_h_l - fil_h_idx * dilation_h;
+      // out_w_tmp = in_w_idx + padding_w_l - fil_w_idx * dilation_w;
+      heightTempIdx = b.create<AffineApplyOp>(
+          loc, heightMap,
+          ValueRange{ivs[genConfig.inputLayout.find('h')], ivs[6]});
+      widthTempIdx = b.create<AffineApplyOp>(
+          loc, widthMap,
+          ValueRange{ivs[genConfig.inputLayout.find('w')], ivs[7]});
+      // out_h_idx = out_h_tmp / stride_h;
+      // out_w_idx = out_w_tmp / stride_w;
+      heightIdx =
+          b.create<arith::FloorDivSIOp>(loc, heightTempIdx, strideHeight);
+      widthIdx = b.create<arith::FloorDivSIOp>(loc, widthTempIdx, strideWidth);
+      break;
+    case miopen::ConvOpType::BwdWeight:
+      // in_h_idx = out_h_idx * stride_h + fil_h_idx * dilation_h - padding_h_l;
+      // in_w_idx = out_w_idx * stride_w + fil_w_idx * dilation_w - padding_w_l;
+      heightIdx = b.create<AffineApplyOp>(
+          loc, heightMap,
+          ValueRange{ivs[6], ivs[genConfig.filterLayout.find('y')]});
+      widthIdx = b.create<AffineApplyOp>(
+          loc, widthMap,
+          ValueRange{ivs[7], ivs[genConfig.filterLayout.find('x')]});
+      break;
+    }
+
+    auto getIndices = [&](std::string tensor) {
```

Comment:
If nothing else, let's go `StringRef`

I'm not liking this stringly-typed approach. Could this be an enum instead?

Reply:
Also, you can take a `SmallVectorImpl<mlir::Value> &` here